src/core/service.c

```diff
@@ -268,38 +268,36 @@ static void service_start_watchdog(Service *s) {
                 log_unit_warning_errno(UNIT(s), r, "Failed to install watchdog timer: %m");
 }
 
-usec_t service_restart_usec(Service *s) {
-        unsigned n_restarts;
+usec_t service_restart_usec_next(Service *s) {
+        unsigned n_restarts_next;
         long double unit;
         usec_t value;
 
         assert(s);
 
-        /* s->n_restarts is not yet updated when we're in these states, so let's add 1 to it manually.
-         * Note that for SERVICE_AUTO_RESTART a restart job might have been enqueued,
-         * i.e. s->n_restarts is already increased. But we assume it's not since the time
-         * between job enqueuing and running is usually neglectable compared to the time
-         * we'll be sleeping. */
-        n_restarts = s->n_restarts +
-                     (IN_SET(s->state, SERVICE_DEAD_BEFORE_AUTO_RESTART, SERVICE_FAILED_BEFORE_AUTO_RESTART, SERVICE_AUTO_RESTART) ? 1 : 0);
+        /* When the service state is in SERVICE_*_BEFORE_AUTO_RESTART or SERVICE_AUTO_RESTART,
+         * we still need to add 1 to s->n_restarts manually because s->n_restarts is not updated
+         * until a restart job is enqueued. Note that for SERVICE_AUTO_RESTART, that might have been
+         * the case, i.e. s->n_restarts is already increased. But we assume it's not since the time
+         * between job enqueuing and running is usually neglectable compared to the time we'll be sleeping. */
+        n_restarts_next = s->n_restarts + 1;
 
-        /* n_restarts can equal to 0 if no restart has happened nor planned */
-        if (n_restarts <= 1 ||
+        if (n_restarts_next == 1 ||
             s->restart_steps == 0 ||
             s->restart_usec_max == USEC_INFINITY ||
             s->restart_usec >= s->restart_usec_max)
                 value = s->restart_usec;
-        else if (n_restarts > s->restart_steps)
+        else if (n_restarts_next > s->restart_steps)
                 value = s->restart_usec_max;
         else {
                 /* Enforced in service_verify() and above */
                 assert(s->restart_usec_max > s->restart_usec);
 
                 unit = powl(s->restart_usec_max - s->restart_usec, 1.0L / s->restart_steps);
-                value = usec_add(s->restart_usec, (usec_t) powl(unit, n_restarts - 1));
+                value = usec_add(s->restart_usec, (usec_t) powl(unit, n_restarts_next - 1));
```

Comment:
If you touch the code, then could you merge the two `powl()`?

Reply:
For readability? I prefer x^yz over (x^y)^z... Or do I misunderstand the implementation?

Reply:
Oh, you meant to calculate the exponent first. Make sense.